Kashoo/synctos
# Issue 301: Support for document expiry

# Description

A document definition may now specify, by way of the `expiry` constraint, when documents of that type will expire, either as a Unix timestamp, ISO 8601 date string, Date object or as a relative time offset in seconds. Includes a new custom action that occurs immediately after document expiry is set: `onExpiryAssignmentSucceeded`.

# Testing

1. Generate a sync function from `expiry-doc-definitions.js`:

```bash
./make-sync-function -j test/resources/expiry-doc-definitions.js build/test-sync-function.js
```

2. Copy and paste the contents of the sync function into a Sync Gateway configuration file.
3. Ensure that the Sync Gateway configuration points to a Couchbase Server bucket and not to a Walrus bucket.
4. Launch Sync Gateway **2.0** with the configuration file.
5. Create a document with a Unix timestamp expiry value. For example:

```
PUT /test/expiry-number-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
	"type": "staticNumberExpiryDoc"
}
```

6. Verify that the `expiry-number-test` document's `expiration` metadata value is set to `4077433455` (i.e. "2099-03-17T12:24:15Z") in the Couchbase Server admin console.
7. Create a document with an ISO 8601 string expiry value. For example:

```
PUT /test/expiry-string-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
	"type": "staticStringExpiryDoc"
}
```

8. Verify that the `expiry-string-test` document's `expiration` metadata value is set to `2556021155` (i.e. "2050-12-30T13:52:35Z") in the Couchbase Server admin console.
9. Create a document with a Date object expiry value. For example:

```
PUT /test/expiry-date-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
	"type": "staticDateExpiryDoc"
}
```

10. Verify that the `expiry-date-test` document's `expiration` metadata value is set to `4269209218` (i.e. "2105-04-15T03:26:58Z") in the Couchbase Server admin console.
11. Create a document with a dynamic numeric offset expiry value. For example:

```
PUT /test/expiry-dynamic-test HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{ 
	"type": "dynamicExpiryDoc",
	"expiry": 3600
}
```

12. Verify that the `expiry-dynamic-test` document's `expiration` metadata value is set to a Unix timestamp that corresponds to one hour in the future in the Couchbase Server admin console. The website https://www.unixtimestamp.com/ might be helpful with this step.

# Related Issue

* GitHub issue link: #301
